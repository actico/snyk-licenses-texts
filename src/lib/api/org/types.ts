export enum SortBy {
  LICENSE = 'license',
  DEPENDENCIES = 'dependencies',
  PROJECTS = 'projects',
  SEVERITY = 'severity',
}

export enum Order {
  ASC = 'asc',
  DESC = 'desc',
}
