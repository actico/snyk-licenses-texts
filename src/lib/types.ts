export interface Dependency {
  id: string; // example: pako@1.0.11
  name: string;
  version: string;
  packageManager: string;
}
//TODO: finish the type
export interface EnrichedDependency extends Dependency {}

export interface LicenseReportDataEntry {
  // TODO: what if it is a dual license?
  /**
   * The text of the license in HTML format
   */
  licenseText: string;
  /**
   * The url of the license
   */
  licenseUrl: string;

  /**
   * License ID
   */
  id: string;
  /**
   * Snyk license severity setup on the org license policy
   */
  severity?: 'none' | 'high' | 'medium' | 'low';
  /**
   * Snyk license instruction setup on the org license policy
   */
  instructions?: string;
  /**
   * Dependencies from this org using this license
   */
  dependencies: DependencyData[];
  /**
   * Snyk projects from this org with dependencies using this license
   */
  projects: {
    id: string;
    name: string;
  }[];
}

export interface DependencyData {
  id: string;
  /**
   * The name of the package
   */
  name: string;
  /**
   * The version of the package
   */
  version: string;
  /**
   * The latest version available for the specified package
   */
  latestVersion?: string;
  /**
   * The timestamp for when the latest version of the specified package was published.
   */
  latestVersionPublishedDate?: string;
  /**
   * The timestamp for when the specified package was first published.
   */
  firstPublishedDate?: string;
  /**
   * True if the latest version of the package is marked as deprecated; False otherwise.
   */
  isDeprecated?: boolean;
  /**
   * The numbers for those versions that are marked as deprecated
   */
  deprecatedVersions?: string[];
  /**
   * The identifiers of dependencies with issues that are depended upon as a result of this dependency
   */
  dependenciesWithIssues?: string[];
  /**
   * The package type of the dependency
   */
  type: string;
  /**
   * The number of high severity issues in this dependency
   */
  issuesHigh?: number;
  /**
   * The number of medium severity issues in this dependency
   */
  issuesMedium?: number;
  /**
   * The number of low severity issues in this dependency
   */
  issuesLow?: number;
  /**
   * The licenses of the dependency
   */
  licenses: {
    /**
     * The identifier of the license
     */
    id: string;
    /**
     * The title of the license
     */
    title: string;
    /**
     * The type of the license
     */
    license: string;
  }[];
  /**
   * The projects which depend on the dependency
   */
  projects: {
    /**
     * The identifier of the project
     */
    id: string;
    /**
     * The name of the project
     */
    name: string;
  }[];
  /**
   * The copyright notices for the package
   */
  copyright?: string[];
}
[];
