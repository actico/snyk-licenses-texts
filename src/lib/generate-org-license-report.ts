import * as debugLib from 'debug';
import * as _ from 'lodash';

export * from './license-text';
export * from './get-api-token';
import { getLicenseDataForOrg, getDependenciesDataForOrg } from './api/org';
import {
  fetchSpdxLicenseTextAndUrl,
  fetchNonSpdxLicenseTextAndUrl,
} from './license-text';
import {
  LicenseReportDataEntry,
  EnrichedDependency,
  Dependency,
  DependencyData,
} from './types';

const debug = debugLib('snyk-licenses:generateOrgLicensesReport');

export interface LicenseReportData {
  [licenseID: string]: LicenseReportDataEntry;
}

export async function generateLicenseData(
  orgPublicId: string,
  options?: {
    filters?: {
      projects: string[];
    };
  },
): Promise<LicenseReportData> {
  debug(`ℹ️  Generating license data for Org:${orgPublicId}`);

  try {
    const licenseData = await getLicenseDataForOrg(orgPublicId, options);
    debug(`✅ Got license API data for Org:${orgPublicId}`);
    const dependenciesDataRaw = await getDependenciesDataForOrg(
      orgPublicId,
      options,
    );
    debug(`✅ Got dependencies API data for Org:${orgPublicId}`);
    const licenseReportData: LicenseReportData = {};
    const dependenciesData = _.groupBy(dependenciesDataRaw.results, 'id');

    if (!licenseData.total) {
      debug(`ℹ️ Detected 0 licenses`);
      return licenseReportData;
    }
    debug(`⏳ Processing ${licenseData.total} licenses`);

    const dependenciesAll = [];

    for (const license of licenseData.results) {
      const dependencies = license.dependencies;
      if (!dependencies.length) {
        continue;
      }
      dependenciesAll.push(...dependencies);
      const dependenciesEnriched = enrichDependencies(
        dependencies,
        dependenciesData,
      );
      if (dependenciesEnriched.length) {
        license.dependencies = dependenciesEnriched;
      }
      const licenseData = await getLicenseTextAndUrl(license.id);
      licenseReportData[license.id] = {
        ...(license as any),
        licenseText: licenseData?.licenseText,
        licenseUrl: licenseData?.licenseUrl,
      };
    }
    debug(`✅ Done processing ${licenseData.total} licenses`);

    return licenseReportData;
  } catch (e) {
    debug('❌ Failed to generate report data', e);
    throw e;
  }
}

function enrichDependencies(
  dependencies: Dependency[],
  dependenciesData,
): EnrichedDependency[] {
  const enrichDependencies: EnrichedDependency[] = [];

  for (const dependency of dependencies) {
    const dep: DependencyData[] = dependenciesData[dependency.id];
    if (dep && dep[0]) {
      enrichDependencies.push({
        ...dependency,
        ...dep[0],
      });
    } else {
      debug('Dep information not available from /dependencies API response for ' + dependency.id);
    }
  }

  return enrichDependencies;
}

async function getLicenseTextAndUrl(
  id: string,
): Promise<{ licenseText: string; licenseUrl: string } | undefined> {
  try {
    return await fetchSpdxLicenseTextAndUrl(id);
  } catch (e) {
    debug(`❌ Failed to get license data for as SPDX, trying non-SPDX: ${id}`);
  }
  try {
    return await fetchNonSpdxLicenseTextAndUrl(id);
  } catch (e) {
    debug(`❌ Failed to get license data as non-SPDX: ${id}`);
  }

  return undefined;
}
