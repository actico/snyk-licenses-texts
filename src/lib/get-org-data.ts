import 'source-map-support/register';
import * as debugLib from 'debug';
import * as snykApiSdk from 'snyk-api-ts-client';
import { getApiToken } from './get-api-token';
import { string } from 'yargs';
import _ = require('lodash');

const debug = debugLib('snyk-licenses:getLicenseDataForOrg');

export interface OrgData {
  name: string;
  id: string;
  slug: string;
  url: string;
  group: {
    name: string;
    id: string;
  };
}

export async function getOrgData(orgPublicId: string): Promise<OrgData> {
  getApiToken();

  try {
    const snykApiClient = await new snykApiSdk.Orgs();
    const allOrgs: OrgData[] = _.get(await snykApiClient.get(), 'orgs', []);
    const orgData = allOrgs.filter(org => org.id === orgPublicId)[0];
    return orgData;
  } catch (e) {
    debug('‚ùå Failed to get org data for org ID:' + orgPublicId);
    throw e;
  }
}
