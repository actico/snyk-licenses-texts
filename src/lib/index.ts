import 'source-map-support/register';
import * as fetch from 'node-fetch';
import * as debugLib from 'debug';
import * as cheerio from 'cheerio';

export async function fetchSpdxLicenseText(licenseId: string): Promise<string> {
  const debug = debugLib('fetchSpdxLicenseText')
  const licenseUrl = `https://spdx.org/licenses/${licenseId}.html`;
  try {
    const res = await fetch(`https://spdx.org/licenses/${licenseId}.html`);
    const rawHtml = await res.text();
    if (res.status !== 200) {
      throw new Error(rawHtml);
    }
    const $ = cheerio.load(rawHtml);
    const licenseText = $('body').text();
    return licenseText;
  } catch (e) {
    debug(`Did not fetch license text successfully. Error: ${e}`);
  }
  return `Did not fetch license text successfully. See the license text at: ${licenseUrl}`;
}
