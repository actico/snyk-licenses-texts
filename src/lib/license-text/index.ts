export { fetchNonSpdxLicenseText } from './non-spdx';
export { fetchSpdxLicenseText } from './spdx';
