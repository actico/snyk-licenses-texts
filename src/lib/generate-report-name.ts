import { OrgData } from "./get-org-data";

export function generateReportName(orgData: OrgData, view: string): string {
  return `${orgData.slug}-${orgData.id}-${view}`;
}
