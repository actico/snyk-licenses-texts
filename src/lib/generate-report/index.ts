import * as Handlebars from 'handlebars';
import * as path from 'path';
import * as fs from 'fs';
import * as debugLib from 'debug';
import * as _ from 'lodash';

import { LicenseReportData } from '../generate-org-license-report';
import { OrgData } from '../get-org-data';
import { SupportedViews, LicenseReportDataEntry } from '../types';

const debug = debugLib('snyk-licenses:generateHtmlReport');
const DEFAULT_TEMPLATE = './templates/licenses-view.hbs';

const transformDataFunc = {
  [SupportedViews.ORG_LICENSES]: transformDataForLicenseView,
  [SupportedViews.PROJECT_DEPENDENCIES]: transformDataForDependencyView,
};

export async function generateHtmlReport(
  orgPublicId: string,
  data: LicenseReportData,
  orgData: OrgData,
  templateOverridePath: string | undefined = undefined,
  view: SupportedViews = SupportedViews.ORG_LICENSES,
): Promise<string> {
  debug('ℹ️  Generating HTML report');
  const hbsTemplate = selectTemplate(view, templateOverridePath);
  debug(
    `✅ Using template ${
      hbsTemplate === DEFAULT_TEMPLATE ? 'default template' : hbsTemplate
    }`,
  );
  debug(`ℹ️ Compiling Handlebars.js template ${hbsTemplate}`);
  const htmlTemplate = await compileTemplate(hbsTemplate);
  debug(`✅ Compiled template ${hbsTemplate}`);
  const transformedData = transformDataFunc[view](orgPublicId, data, orgData);
  return htmlTemplate(transformedData);
}

function transformDataForLicenseView(
  orgPublicId: string,
  data: LicenseReportData,
  orgData: OrgData,
): {
  licenses: LicenseReportData;
  orgPublicId: string;
  orgData: OrgData;
} {
  return { licenses: data, orgPublicId, orgData };
}

interface ProjectsReportData {
  [projectId: string]: {
    projectName: string;
    projectIndex: number;
    projectId: string;
    licenses: {
      [licenseId: string]: LicenseReportDataEntry;
    };
  };
}

function transformDataForDependencyView(
  orgPublicId: string,
  data: LicenseReportData,
  orgData: OrgData,
): {
  projects: ProjectsReportData;
  orgPublicId: string;
  orgData: OrgData;
  totalProjects: number;
} {
  const projectData: ProjectsReportData = {};
  let totalProjects = 0;

  for (const licenseId of Object.keys(data)) {
    const licenseData = data[licenseId];

    for (const project of licenseData.projects) {
      if (!projectData[project.id]) {
        totalProjects += 1;
        projectData[project.id] = {
          projectIndex: totalProjects,
          projectId: project.id,
          projectName: project.name,
          licenses: {
            [licenseId]: licenseData,
          },
        };
      } else {
        projectData[project.id].licenses[licenseId] = licenseData;
      }
    }
  }

  return {
    projects: projectData,
    orgPublicId,
    orgData,
    totalProjects,
  };
}

function selectTemplate(view: SupportedViews, templateOverride?): string {
  switch (view) {
    case SupportedViews.ORG_LICENSES:
      return templateOverride || DEFAULT_TEMPLATE;
    case SupportedViews.PROJECT_DEPENDENCIES:
      return templateOverride || './templates/project-dependencies-view.hbs';
    default:
      return DEFAULT_TEMPLATE;
  }
}

async function compileTemplate(
  fileName: string,
): Promise<HandlebarsTemplateDelegate> {
  return readFile(path.resolve(__dirname, fileName), 'utf8').then(
    Handlebars.compile,
  );
}

function readFile(filePath: string, encoding: string): Promise<string> {
  return new Promise<string>((resolve, reject) => {
    fs.readFile(filePath, encoding, (err, data) => {
      if (err) {
        reject(err);
      }
      resolve(data);
    });
  });
}
