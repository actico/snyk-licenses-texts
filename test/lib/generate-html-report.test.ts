import * as path from 'path';
import { generateHtmlReport } from '../../src/lib/generate-report';
import { loadJson } from '../load-json';
import { LicenseReportData } from '../../src/lib/generate-org-license-report';
import { SupportedViews } from '../../src/lib/types';

describe('Generate HTML report', () => {
  const OLD_ENV = process.env;
  process.env.SNYK_TOKEN = process.env.SNYK_TEST_TOKEN;
  const ORG_ID = process.env.TEST_ORG_ID as string;

  afterAll(async () => {
    process.env = { ...OLD_ENV };
  });
  test('SNYK_TOKEN & ORG_ID are set', async () => {
    expect(process.env.SNYK_TOKEN).not.toBeNull();
    expect(process.env.ORG_ID).not.toBeNull();
  });
  test('License HTML Report is generated as expected', async () => {
    const licenseRes = (loadJson(
      __dirname + '/fixtures/example-license-data.json',
    ) as unknown) as LicenseReportData;
    const orgData = {
      name: 'org',
      id: 'avd-scv',
      slug: 'org',
      url: 'https://snyk.io/org/org',
      group: {
        name: 'group',
        id: 'group-1',
      },
    };
    const htmlData = await generateHtmlReport(ORG_ID, licenseRes, orgData);
    expect(htmlData).toMatchSnapshot();
  }, 50000);

  test('License HTML Report is generated as expected with project based view', async () => {
    const licenseRes = (loadJson(
      __dirname + '/fixtures/example-license-data.json',
    ) as unknown) as LicenseReportData;
    const orgData = {
      name: 'org',
      id: 'avd-scv',
      slug: 'org',
      url: 'https://snyk.io/org/org',
      group: {
        name: 'group',
        id: 'group-1',
      },
    };
    const htmlData = await generateHtmlReport(
      ORG_ID,
      licenseRes,
      orgData,
      undefined,
      SupportedViews.PROJECT_DEPENDENCIES,
    );
    expect(htmlData).toMatchSnapshot();
  }, 50000);

  test('License HTML Report is generated as expected with a custom hbs template', async () => {
    const licenseRes = (loadJson(
      __dirname + '/fixtures/example-license-data.json',
    ) as unknown) as LicenseReportData;
    const orgData = {
      name: 'org',
      id: 'avd-scv',
      slug: 'org',
      url: 'https://snyk.io/org/org',
      group: {
        name: 'group',
        id: 'group-1',
      },
    };
    const htmlData = await generateHtmlReport(
      ORG_ID,
      licenseRes,
      orgData,
      path.resolve(__dirname + '/fixtures/custom-view.hbs'),
    );
    expect(htmlData).toMatchSnapshot();
  }, 50000);
  test.todo('Test for when API fails aka bad org id provided');
});
