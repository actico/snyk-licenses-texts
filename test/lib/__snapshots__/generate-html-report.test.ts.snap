// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate HTML report License HTML Report is generated as expected 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <meta http-equiv=\\"Content-type\\" content=\\"text/html; charset=utf-8\\">
    <meta http-equiv=\\"Content-Language\\" content=\\"en-us\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
    <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
    <title>Snyk Licenses Report</title>
    <link rel=\\"icon\\" type=\\"image/png\\" href=\\"https://res.cloudinary.com/snyk/image/upload/v1468845142/favicon/favicon.png\\"
      sizes=\\"194x194\\">
    <link rel=\\"shortcut icon\\" href=\\"https://res.cloudinary.com/snyk/image/upload/v1468845142/favicon/favicon.ico\\">
  </head>

  <body>
    <main class=\\"layout-stacked\\">
      <h1>Hello!</h1>
    </main>
  </body>
</html>

"
`;
